Document `TRACK_DIAGNOSTIC` calls.

nnethercote opened the pull request:
r? @cjgillot

nnethercote commented:
@cjgillot: I'm fairly confident this is correct, but not 100%. I'm hoping you can confirm it.

@bors rollup=always

bors commented:
:umbrella: The latest upstream changes (presumably #120715) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote commented:
It's a tiny change, but I still managed to make this conflict with one of my own PRs :)

Anyway, I rebased.

nnethercote committed:
Give `TRACK_DIAGNOSTIC` a return value.

This means `DiagCtxtInner::emit_diagnostic` can return its result
directly, rather than having to modify a local variable.

nnethercote commented:
I added a small follow-up refactoring.

cjgillot commented on compiler/rustc_errors/src/lib.rs:
```diff
@@ -516,12 +516,16 @@ pub enum StashKey {
 
-pub static TRACK_DIAGNOSTIC: AtomicRef<fn(Diagnostic, &mut dyn FnMut(Diagnostic))> =
-    AtomicRef::new(&(default_track_diagnostic as _));
+pub static TRACK_DIAGNOSTIC: AtomicRef<
```
Could you add a doc-comment?

cjgillot commented on compiler/rustc_errors/src/lib.rs:
```diff
@@ -1295,20 +1299,15 @@ impl DiagCtxtInner {
 
-        let mut guaranteed = None;
-        (*TRACK_DIAGNOSTIC)(diagnostic, &mut |mut diagnostic| {
+        TRACK_DIAGNOSTIC(diagnostic, &mut |mut diagnostic| {
```
Could you add a comment on this function explaining that we only need to track diagnostics that are actually emitted?

cjgillot commented on compiler/rustc_errors/src/lib.rs:
```diff
@@ -1295,20 +1299,15 @@ impl DiagCtxtInner {
             }
             Allow | Expect(_) => {
-                (*TRACK_DIAGNOSTIC)(diagnostic, &mut |_| {});
                 return None;
```
Is this correct for `Expect` diagnostics too?
The second execution is not really a no-op, as this function will add a side-effect to `self.fulfilled_expectations`.
I'm not certain that we correctly test this path in incremental.

nnethercote replied:
Hmm, I think you are right. In fact, similar reasoning applies for the `Warning` case -- there is the side-effect of `self.future_breakage_diagnostics` being updated on the replay. So neither call can be removed.

In fact, now I suspect there is a missing `TRACK_DIAGNOSTIC` call for this code:
```
            if let LintExpectationId::Unstable { .. } = expectation_id {
                self.unstable_expect_diagnostics.push(diagnostic);
                return None;
            }
```
because there's the side-effect of updating `self.unstable_expect_diagnostics`.

nnethercote replied:
And now I'm wondering if there's a missing `TRACK_DIAGNOSTIC` for the delayed bugs as well :(

nnethercote replied:
Are you sure about the delayed bugs case? They don't doom compilation unless no other errors are emitted. So if we do diagnostic replay and don't reply delayed bugs, I think we are possibly missing out on some internal sanity checking.

cjgillot reviewed (commented):
I took a bit of time before getting to this. I think the reasoning is correct: we only need to track actual side effects, not potential side effects. However, I have a doubt about the `Expect` case.

nnethercote committed:
Inline and remove `Level::get_diagnostic_id`.

It has a single call site, and this will enable subsequent refactorings.

nnethercote committed:
Move `DelayedBug` handling into the `match`.

It results in a tiny bit of duplication (another
`self.treat_next_err_as_bug()` condition) but I think it's worth it to
get more code into the main `match`.

nnethercote committed:
Reorder `has_future_breakage` handling.

This will enable additional refactorings.

nnethercote committed:
Move `Expect`/`ForceWarning` handling into the `match`.

Note that `self.suppressed_expected_diag` is no longer set for
`ForceWarning`, which is good. Nor is `TRACK_DIAGNOSTIC` called for
`Allow`, which is also good.

nnethercote committed:
Add comments about `TRACK_DIAGNOSTIC` use.

Also add an assertion for the levels allowed with `has_future_breakage`.

nnethercote commented:
I have redone this PR. There's a bunch of refactoring, plus some addition of comments. Quite different from the original version, but I think it adds some welcome clarity to `emit_diagnostic`.

cjgillot commented on compiler/rustc_errors/src/lib.rs:
```diff
@@ -1331,6 +1333,10 @@ impl DiagCtxtInner {
+                        // Is saving the diagnostic in `delayed_bugs` a notable
+                        // side-effect? Should `TRACK_DIAGNOSTIC` be called?
+                        // Unclear. Currently we err on the side of "no" to avoid
+                        // having to clone the diagnostic.
```
We do not need to call `TRACK_DIAGNOSTIC`: if there is a delayed bug, the incremental session is deleted.

nnethercote replied:
Ok, I have updated the comment.

nnethercote replied:
:+1:

bors commented:
:umbrella: The latest upstream changes (presumably #121415) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

nnethercote committed:
Make the `match` in `emit_diagnostic` complete.

This match is complex enough that it's a good idea to enumerate every
variant.

This also means `can_be_top_or_sub` can just be `can_be_subdiag`.

nnethercote commented:
I rebased and added one new commit, to make the `match` in `emit_diagnostic` enumerate every variant. I think this is good to merge now, I like how much it has clarified all the different cases in `emit_diagnostic`.

nnethercote commented:
@cjgillot has gone on vacation for two weeks, so I will reassign:

r? @oli-obk

We have already gone through one round of review feedback and updates, I think @cjgillot was close to approving it.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 7ef605be3fa24f93194a3faf4f75d3a05ceca78a has been approved by `oli-obk`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=oli-obk 7ef605be3fa24f93194a3faf4f75d3a05ceca78a -->
<!-- homu: {"type":"Approved","sha":"7ef605be3fa24f93194a3faf4f75d3a05ceca78a","approver":"oli-obk","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
